lib/src/front_end/comment_writer.dart

```diff
@@ -0,0 +1,277 @@
+      var comment = comments[i];
+      var containsNewline =
+          comment.type == CommentType.line || comment.text.contains('\n');
+
+      if (_isHangingComment(comment)) {
+        // Attach the comment to the previous token.
+        writer.space();
+        writer.write(comment.text,
+            containsNewline: containsNewline, following: true);
+      } else {
+        writer.writeNewline();
+        writer.write(comment.text, containsNewline: containsNewline);
+      }
+
+      if (comment.type == CommentType.line) writer.writeNewline();
+    }
+
+    if (comments.isNotEmpty && _needsSpaceAfterComment(token.lexeme)) {
+      writer.space();
+    }
+  }
+
+  /// Takes all of the comment tokens preceding [token] and converts them to
+  /// [SourceComment]s that track their kind and the whitespace between them.
+  ///
+  /// Returns the list of [SourceComments] and the number of newlines between
+  /// the last comment and [token]. If there are no comments, returns an empty
+  /// list and the number of lines between [token] and the preceding token.
+  (List<SourceComment>, int) _convertComments(Token token) {
+    Token? comment = token.precedingComments;
+
+    // TODO(perf): Avoid calculating newlines between tokens unless there are
+    // comments or it's needed to determine whether to insert a blank in a
+    // sequence.
+
+    // TODO(tall): If the token's comments are being moved by a fix, do not
+    // write them here.
+
+    var previousLine = _endLine(token.previous!);
+    var tokenLine = _startLine(token);
+
+    // Edge case: The analyzer includes the "\n" in the script tag's lexeme,
+    // which confuses some of these calculations. We don't want to allow a
+    // blank line between the script tag and a following comment anyway, so
+    // just override the script tag's line.
+    if (token.previous!.type == TokenType.SCRIPT_TAG) previousLine = tokenLine;
+
+    var comments = <SourceComment>[];
+    while (comment != null) {
+      var commentLine = _startLine(comment);
+
+      // Don't preserve newlines at the top of the file.
+      if (comment == token.precedingComments &&
+          token.previous!.type == TokenType.EOF) {
+        previousLine = commentLine;
+      }
+
+      var text = comment.lexeme.trim();
+      var linesBefore = commentLine - previousLine;
+      var flushLeft = _startColumn(comment) == 1;
+
+      if (text.startsWith('///') && !text.startsWith('////')) {
+        // Line doc comments are always indented even if they were flush left.
+        flushLeft = false;
+
+        // Always add a blank line (if possible) before a doc comment block.
+        if (comment == token.precedingComments) linesBefore = 2;
+      }
+
+      var type = CommentType.block;
+      if (text.startsWith('///') && !text.startsWith('////') ||
+          text.startsWith('/**') && text != '/**/') {
+        // TODO(rnystrom): Check that the comment isn't '/**/' because some of
```

Comment:
Why is this a TODO - what do we expect to change here? The current code looks correct to me

Reply:
Good point. Updated the comment.